src/effects/effectchainslot.cpp

```diff
@@ -114,12 +144,408 @@ EffectChainSlot::~EffectChainSlot() {
+    // not needed.
+    if (!m_bAddedToEngine || bWasAlreadyEnabled) {
+        return;
+    }
+
+    EffectsRequest* request = new EffectsRequest();
+    request->type = EffectsRequest::ENABLE_EFFECT_CHAIN_FOR_INPUT_CHANNEL;
+    request->pTargetChain = m_pEngineEffectChain;
+    request->EnableInputChannelForChain.pChannelHandle = &handle_group.handle();
+
+    // Allocate EffectStates here in the main thread to avoid allocating
+    // memory in the realtime audio callback thread. Pointers to the
+    // EffectStates are passed to the EffectRequest and the EffectProcessorImpls
+    // store the pointers. The containers of EffectState* pointers get deleted
+    // by ~EffectsRequest, but the EffectStates are managed by EffectProcessorImpl.
+    auto pEffectStatesMapArray = new EffectStatesMapArray;
+
+    //TODO: get actual configuration of engine
+    const mixxx::EngineParameters bufferParameters(
+          mixxx::AudioSignal::SampleRate(96000),
+          MAX_BUFFER_LEN / mixxx::kEngineChannelCount);
+
+    for (int i = 0; i < m_effects.size(); ++i) {
+        auto& statesMap = (*pEffectStatesMapArray)[i];
+        if (m_effects[i] != nullptr) {
+            for (const auto& outputChannel : m_pEffectsManager->registeredOutputChannels()) {
+                if (kEffectDebugOutput) {
+                    qDebug() << debugString() << "EffectChain::enableForInputChannel creating EffectState for input" << handle_group << "output" << outputChannel;
+                }
+                statesMap.insert(outputChannel.handle(),
+                        m_effects[i]->createState(bufferParameters));
+            }
+        } else {
+            for (EffectState* pState : statesMap) {
+                if (pState != nullptr) {
+                    delete pState;
+                }
+            }
+            statesMap.clear();
+        }
+    }
+    request->EnableInputChannelForChain.pEffectStatesMapArray = pEffectStatesMapArray;
+
+    m_pEffectsManager->writeRequest(request);
+    // emit(channelStatusChanged(handle_group.name(), true));
+    slotChainChannelStatusChanged(handle_group.name(), true);
+}
+
+bool EffectChainSlot::enabledForChannel(const ChannelHandleAndGroup& handle_group) const {
+    return m_enabledInputChannels.contains(handle_group);
+}
+
+void EffectChainSlot::disableForInputChannel(const ChannelHandleAndGroup& handle_group) {
+    if (m_enabledInputChannels.remove(handle_group)) {
+        if (!m_bAddedToEngine) {
+            return;
+        }
+        EffectsRequest* request = new EffectsRequest();
+        request->type = EffectsRequest::DISABLE_EFFECT_CHAIN_FOR_INPUT_CHANNEL;
+        request->pTargetChain = m_pEngineEffectChain;
+        request->DisableInputChannelForChain.pChannelHandle = &handle_group.handle();
+        m_pEffectsManager->writeRequest(request);
+
+        // emit(channelStatusChanged(handle_group.name(), false));
+        slotChainChannelStatusChanged(handle_group.name(), false);
```

Comment:
I think this code follows a circular path now. All that `slotChainChannelStatusChanged` does is update EffectChainSlot's ChannelInfo, but `enableForInputChannel` and `disableForInputChannel` should only be called by `slotChannelStatusChanged` when the ChannelInfo's ControlObject emits its `valueChanged` signal.